File: Solution/CSharp/160.IntersectionofTwoLinkedLists.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode GetIntersectionNode(ListNode headA, ListNode headB) {
        int countA = GetCount(headA);
        int countB = GetCount(headB);
        int offset = Math.Abs(countA - countB);

        if (countA > countB) {
            headA = GetIndex(headA, offset);
            headB = GetIndex(headB, offset);
        }

        while (headA != null && headB != null) {
            if (headA.GetHashCode() == headB.GetHashCode()) {
                return headA;
            } else {
                headA = headA.next;
                headB = headB.next;
            }
        }
        return null;
    }

    public ListNode GetIndex(ListNode head, int index) {
        int count = 0;
        ListNode node = head;
        if (count == index) {
            return node;
        } else {
            count++;
            node = node.next;
        }
    }
    return null;
}

public int GetCount(ListNode head) {
    int count = 0;
    ListNode node = head;
    while (node != null) {
        count++;
        node = node.next;
    }
    return count;
}
}

File: Solution/Python/160. Intersection of Two Linked Lists.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if headA == None and headB == None:
            return None
        pa = headA
        pb = headB
        while pa != pb:
            pa = headB if pa == None else pa.next
            pb = headA if pb == None else pb.next
        return None


class Solution2(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if not headA or not headB:
            return None
        countA = countB = 0

        nodeA = headA
        nodeB = headB
        while nodeA or nodeB:
            if nodeA:
                countA += 1
                nodeA = nodeA.next
            if nodeB:
                countB += 1
                nodeB = nodeB.next

        nodeA = headA
        nodeB = headB
        while countA != countB:
            if countA > countB:
                nodeA = nodeA.next
                countA -= 1
            elif countA < countB:
                countB -= 1
                nodeB = nodeB.next

        while nodeA and nodeA != nodeB:
            nodeA = nodeA.next
            nodeB = nodeB.next

        return nodeA


c = ListNode(10)
c.next = ListNode(11)
c.next.next = ListNode(12)

a = ListNode(1)
a.next = ListNode(2)
a.next.next = c

b = ListNode(1)
b.next = ListNode(2)
b.next.next = ListNode(3)
b.next.next.next = c

res = Solution2().getIntersectionNode(a, b)
print(res.val)
